### PR TITLE
Added an implementation of a multivariate normal distribution

### PR DESCRIPTION
**Pull request by tbrx**
This also includes updates for the matrix constraints in `distributions.constraints` to handle batches of matrices.

Linear algebra operations fall back to python list comprehensions and `torch.stack` in a few cases, where batched versions are not yet available. In particular, this would benefit from a batched implementation of `torch.potrf` and a batched and differentiable `torch.potrs` or `torch.trtrs`.

### CODE REVIEW

File: torch/distributions/constraints.py
Comment (apaszke) [inline torch/distributions/constraints.py]: That should be unnecessary. `torch.tril` will slice out only the part above the diagonal from `value`

File: torch/distributions/constraints.py
Comment (apaszke) [inline torch/distributions/constraints.py]: Same here. Why did that change?

File: torch/distributions/constraints.py
Comment (apaszke) [inline torch/distributions/constraints.py]: NB it would be faster to flatten last two dims (you don't need contiguous because `==` will always give you a contig tensor) and call `min` once

File: torch/distributions/multivariate_normal.py
Comment (apaszke) [inline torch/distributions/multivariate_normal.py]: Why do we need all this broadcasting here? Doesn't `torch.bmm` do it?

File: torch/distributions/multivariate_normal.py
Comment (apaszke) [inline torch/distributions/multivariate_normal.py]: Wouldn't this work (`size + 1` trick is a nice way to get the diagonal without any copying):
```python
bmat.view(bmat.size(0), -1)[:, ::bmat.size(1) + 1]
```

File: torch/distributions/multivariate_normal.py
Comment (apaszke) [inline torch/distributions/multivariate_normal.py]: It looks like we're using `scale_tril` to implement pretty much all of the operations here, so it would be nice to add a note in the docs, saying that it's preferable to provide this arg.

**Review by apaszke — COMMENTED** [review]
Looks good for the most part, but the linter has a lot of complaints. Has the math in this PR been reviewed in probtorch?

File: torch/distributions/constraints.py
Comment (tbrx) [reply torch/distributions/constraints.py]: Does that work on batches of matrices? If I call `torch.tril(torch.randn(3,2,2))` then I get a `RuntimeError: invalid argument 1: expected a matrix`…

**Comment by tbrx** [comment]
Thanks @apaszke — the previous discussion of this (implementation choices, math) is over at probtorch/pytorch#52.

File: torch/distributions/multivariate_normal.py
Comment (tbrx) [reply torch/distributions/multivariate_normal.py]: To call `torch.bmm`, as I understand it we need both arguments to be 3D tensors. If the batch_shape is multidimensional, then this flattens it down. It's more complicated here than elsewhere because the matrix `bmat` might be missing leading dimensions of the batch shape.

I wanted to avoid pre-emptively calling `.expand` on the `scale_tril` matrix, since in the `_batch_mahalanobis` helper (line 71) we split along the batch dimension, call `torch.inverse`, and then call `torch.stack`. 

This whole problem comes about because the leading dimensions of the`loc` and `scale_tril` parameters might not be the same. If the batch size is large (e.g. because loc has a batch dimension) but there is only a single covariance matrix, then we don't mistakenly end up inverting the single covariance matrix once for every item in the batch. (Originally we had also discussed prohibiting any batching of the `covariance_matrix` or `scale_tril` parameter.)

If I understand correctly, I think we could probably remove this entire workaround after #4612.

File: torch/distributions/multivariate_normal.py
Comment (tbrx) [reply torch/distributions/multivariate_normal.py]: That's totally awesome. Updated.

File: torch/distributions/constraints.py
Comment (apaszke) [inline torch/distributions/constraints.py]: I now see the problem with 3D inputs, but can we please make batch tril a helper in utils that:
1. just calls tril on 2D inputs
2. does this workaround with multiplication otherwise
I'd rather do it that way, because the workaround makes it slower in the 2d case

File: torch/distributions/constraints.py
Comment (apaszke) [inline torch/distributions/constraints.py]: nit: can you please add a comment that it fails for NaNs (it's not always obvious that `x == x` can ever be false)

File: torch/distributions/multivariate_normal.py
Comment (apaszke) [inline torch/distributions/multivariate_normal.py]: I think it should be possible to replace view + expand by `.expand(bvec.shape[0], event_dim, event_dim)`

File: torch/distributions/multivariate_normal.py
Comment (apaszke) [reply torch/distributions/multivariate_normal.py]: Can you please document what are the expected sizes of inputs to this function? It's very hard to review it without checking and fully understanding all the places where it's used.

**Review by apaszke — COMMENTED** [review]
LGTM. Some minor comments and it should be good to land

File: torch/distributions/multivariate_normal.py
Comment (tbrx) [reply torch/distributions/multivariate_normal.py]: Apologies — this was really far more complicated than it needed to be. I rewrote (and documented) this helper, and it should be much more clear now.

**Review by apaszke — APPROVED** [review]
Looks great now!

**Comment by apaszke** [comment]
Can you please resolve the conflicts? Will be ready to merge afterwards

**Comment by tbrx** [comment]
@vishwakftw Do you know why I would be getting a test failure for one of the `TransformedDistribution` examples in `test_cdf_log_prob` after merging in master…? It looks like a sign is off somehow. I don't think I should have touched anything that affects this test.

```
======================================================================
FAIL: test_cdf_log_prob (__main__.TestDistributions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_distributions.py", line 1325, in test_cdf_log_prob
    'grad(cdf) = {}'.format(cdfs_derivative),
  File "/Users/brookspaige/Documents/pytorch/pytorch/test/common.py", line 233, in assertEqual
    assertTensorsEqual(x, y)
  File "/Users/brookspaige/Documents/pytorch/pytorch/test/common.py", line 225, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: TransformedDistribution example 3/3, d(cdf)/dx != pdf(x)
x = Variable containing:
 0.1667  0.2383  0.1121
 0.2098  0.2094  0.1962
[torch.DoubleTensor of size (2,3)]

cdf = Variable containing:
 0.9476  0.0065  0.7071
 0.0285  0.7694  0.1535
[torch.DoubleTensor of size (2,3)]

pdf = Variable containing:
 6.9085  0.7263  2.6612
 0.5197  2.3816  2.2812
[torch.DoubleTensor of size (2,3)]

grad(cdf) = Variable containing:
-6.9085 -0.7263 -2.6612
-0.5197 -2.3816 -2.2812
[torch.DoubleTensor of size (2,3)]
```

**Comment by vishwakftw** [comment]
@tbrx I have no idea why the tests are failing, especially due to a polarity difference. I checked the `TransformedDistribution`'s `.cdf` function, `Normal`'s `.cdf` function and could not find any mistakes w.r.t. them. It is also weird since this error didn't turn up when this got merged into master as well.

cc: @fritzo

File: test/test_distributions.py
Comment (vishwakftw) [inline test/test_distributions.py]: Why did this not affect any earlier tests?

**Comment by fritzo** [comment]
Ok I just pushed a one-line workaround to the failed test. The reason if failed in this PR is that @tbrx inserted a `MultivariateNormal` in `EXAMPLES` above `TransformedDistribution`, thereby changing the randomness in that example.

There is a real issue in `TransformedDistribution.cdf()` and `.icdf()` where an assumption is made that the transform is monotone increasing rather than monotone decreasing. This came up in this PR because an `AffineTransform` set `scale < 0`. I'm not sure what is the right thing to do here, i.e. whether there is a generic method to determine whether a transform is monotone increasing or decreasing. https://github.com/probtorch/pytorch/issues/127

**Comment by vishwakftw** [comment]
@tbrx I think the fix my @fritzo needs to be reverted.

File: torch/distributions/multivariate_normal.py
Comment (vishwakftw) [inline torch/distributions/multivariate_normal.py]: Considering the recent merging of Tensor and Variable classes, could you please update this?

File: torch/distributions/multivariate_normal.py
Comment (vishwakftw) [inline torch/distributions/multivariate_normal.py]: Should it be mentioned `using Cholesky decomposition`?

File: torch/distributions/multivariate_normal.py
Comment (vishwakftw) [inline torch/distributions/multivariate_normal.py]: Is it always two-dimensional or a batch of two-dimensional matrices?

File: torch/distributions/constraints.py
Comment (vishwakftw) [inline torch/distributions/constraints.py]: Is it safe to keep the greater than 0.0 check as:
```python
not (v.symeig()[0][:1]).abs() < 1e-12
```
?

File: torch/distributions/multivariate_normal.py
Comment (tbrx) [reply torch/distributions/multivariate_normal.py]: Could you clarify how exactly it should be updated? I'm a bit confused whether the new combined classes should be referred to as `Tensor` or as `Variable`. I tried looking for the updated docs, but I don't believe they've been merged into master yet.

Should we also be updating all example and test code to avoid the `Variable` wrapper?

Comment (apaszke) [reply torch/distributions/multivariate_normal.py]: Variable will soon disappear and be replaced with Tensor

File: torch/distributions/multivariate_normal.py
Comment (tbrx) [reply torch/distributions/multivariate_normal.py]: Do you have a suggestion how we should word this? Tentatively updating to:

"scale_tril matrix must be two-dimensional, with optional leading batch dimensions"

File: torch/distributions/constraints.py
Comment (vishwakftw) [reply torch/distributions/constraints.py]: This is an [example](https://drive.google.com/file/d/1NyyMX7YJz4nJlNxOzyrO6mAZVwB2gVwv/view?usp=sharing) (dumped using `numpy.savetxt`, so you could load using `numpy.loadtxt` or `numpy.genfromtxt`). These are the top 10 eigenvalues for that matrix:
```python
9.96764812e-01 +0.00000000e+00j,
1.42575611e-10 +0.00000000e+00j,
-1.05092455e-10 +8.56060079e-11j,
-1.05092455e-10 -8.56060079e-11j,
-8.01922082e-11 +5.76392912e-11j,
-8.01922082e-11 -5.76392912e-11j,
9.43347083e-11 +1.54831723e-11j,
9.43347083e-11 -1.54831723e-11j,
7.89421910e-11 +1.71497236e-11j,
7.89421910e-11 -1.71497236e-11j
```

**Review by apaszke — APPROVED** [review]
Applied a very minor patch. LGTM.

File: torch/distributions/constraints.py
Comment (tbrx) [reply torch/distributions/constraints.py]: That matrix is not actually symmetric, so it should also fail this constraint check anyway. Probably we should add an explicit check that the matrix is symmetric as well — I hadn't thought of that (I had figured, incorrectly and without checking, that `torch.symeig` would barf if given an assymmetric matrix).

Comment (apaszke) [reply torch/distributions/constraints.py]: Do you guys want to still work on this, or should I just go ahead and merge this PR?

**Comment by fritzo** [comment]
Is anything blocking this PR? Does it need to be updated to use new `torch.tensor` idioms?

**Comment by tbrx** [comment]
@fritzo Apparently so — many of the tests are now failing. Will update today.

**Comment by fritzo** [comment]
Yay, glad to finally see this merged!

**Pull request merged**
